src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -136,6 +138,36 @@ internal SourceText ReadFileContent(CommandLineSourceFile file, IList<Diagnostic
             return ReadFileContent(file, diagnostics, out discarded);
         }
 
+        internal EmbeddedText ReadEmbeddedFileContent(CommandLineSourceFile file, IList<DiagnosticInfo> diagnostics)
+        {
+            try
+            {
+                // PERF: Using a very small buffer size for the FileStream opens up an optimization within EncodedStringText where
+                // we read the entire FileStream into a byte array in one shot. For files that are actually smaller than the buffer
+                // size, FileStream.Read still allocates the internal buffer.
+                using (var stream = PortableShim.FileStream.Create(file.Path, PortableShim.FileMode.Open, PortableShim.FileAccess.Read, PortableShim.FileShare.ReadWrite, bufferSize: 1, options: PortableShim.FileOptions.None))
+                {
+                    const int LargeObjectHeapLimit = 80 * 1024;
+
+                    if (stream.Length < LargeObjectHeapLimit)
+                    {
+                        byte[] buffer = EncodedStringText.TryGetByteArrayFromStream(stream);
+                        if (buffer != null)
+                        {
+                            return EmbeddedText.FromBytes(file.Path, buffer, (int)stream.Length, Arguments.ChecksumAlgorithm);
+                        }
+                    }
+
+                    return EmbeddedText.FromStream(file.Path, stream, Arguments.ChecksumAlgorithm);
+                }
```

Comment:
Looks familiar. Don't we use this pattern elsewhere? Could we generalize to a helper?